lib/createNavLink.js

```diff
@@ -0,0 +1,176 @@
+/**
+ * Copyright 2015, Yahoo! Inc.
+ * Copyrights licensed under the New BSD License. See the accompanying LICENSE file for terms.
+ */
+/*global window */
+'use strict';
+var React = require('react');
+var RouteStore = require('./RouteStore');
+var debug = require('debug')('NavLink');
+var navigateAction = require('./navigateAction');
+var objectAssign = require('object-assign');
+
+function isLeftClickEvent (e) {
+    return e.button === 0;
+}
+
+function isModifiedEvent (e) {
+    return !!(e.metaKey || e.altKey || e.ctrlKey || e.shiftKey);
+}
+
+/**
+ * create NavLink with custom options
+ * @param {Object} options to overwrite the default attributes to create NavLink
+ * @returns {React.Component} NavLink
+ */
+module.exports = function createNavLink (options) {
+    var NavLink = React.createClass(objectAssign({}, {
+        displayName: 'NavLink',
+        contextTypes: {
+            executeAction: React.PropTypes.func,
+            getStore: React.PropTypes.func
+        },
+        propTypes: {
+            currentRoute: React.PropTypes.object,
+            currentNavigate: React.PropTypes.object,
+            href: React.PropTypes.string,
+            stopPropagation: React.PropTypes.bool,
+            routeName: React.PropTypes.string,
+            navParams: React.PropTypes.object,
+            followLink: React.PropTypes.bool,
+            preserveScrollPosition: React.PropTypes.bool,
+            replaceState: React.PropTypes.bool
+        },
+        getInitialState: function () {
+            return this._getState(this.props);
+        },
+        componentDidMount: function () {
+            var routeStore = this.context.getStore(RouteStore);
+            routeStore.addChangeListener(this._onRouteStoreChange);
+        },
+        componentWillUnmount: function () {
+            var routeStore = this.context.getStore(RouteStore);
+            routeStore.removeChangeListener(this._onRouteStoreChange);
+        },
+        shouldComponentUpdate: function (nextProps, nextState) {
+            if (this.state.isActive !== nextState.isActive || this.receivedNewProps) {
+                return true;
+            }
+            return false;
+        },
+        componentWillReceiveProps: function (nextProps) {
+            this.receivedNewProps = true;
```

Comment:
We want to disable the re-render for route store change, (only render when isActive change), while we still need to re-render for props change, so set a flag `receivedNewProps` and clear it after rendered.